_posts/2024/03/2024-03-18-cage-recursion.md

```diff
@@ -0,0 +1,100 @@
+---
+layout: post
+date: 2024-03-18
+title: "Recursion in cage"
+author: levBagryansky
+---
+
+The main difference between `cage` and `memory` is how the object is written.
+When dataizing, `memory.write` executes its argument as well. Whereas 
+`cage.write` just saves a reference to this object. As a result, the execution
+of the `cage` may turn out to be recursive. Unfortunately, this error is often not obvious.
+
+In this post we will tell you how we solved this problem.
+
+<!--more-->
+
+# A little about cage
+
+In [this](../../2023/08/2023-08-05-storing-objects-formed-differently-into-cage.md) post we 
+raised the topic of cage. There we told how to write into cage, and what limitations there are -
+objects of the same `forma` can be written into one cage.
+Here is a simple example of using a cage:
+
+```eo
+# Cage conatains int.
+[] > cage-contains-int
+  cage 0 > my-int
+  eq. > @
+    seq
+      *
+        my-int.write 42
+        my-int
+        sum
+    42
+```
+
+But you might also accidentally write cage inside cage, leading to infinite recursion.
+```eo
+[] > catches-stack-overflow
+  cage > cge
+    0.plus
+  seq > @
+    *
+      cge.write
+        0.plus cge
+      cge
+```
+In order to provide everywhere the same `forma` we use `0.plus`.
+When `cge` is dataizing it dataizes `0.plus cge`. It takes attribute `plus` from object `0`. `int.plus`
+tries to dataize its argument `cge` inside. We have achieved our goal: infinite recursion is obtained.
+The first attempt is to detect such cases even in compilation. But generally speaking, the presence 
+of `case` inside `case` is not a mistake, because such a program may well be correct.
+Dataization of cage can several times pas through the `cage` and only then return some value.
+This eo illustrates such scenario (it does not work with latest version but idea of deep cage is clear):
+```eo
+# correct-cage-recursion.
+[] > correct-cage-recursion
+  cage > cge
+    if.
+      i.as-bool
+      TRUE
+      seq
+        *
+          i.write
+            TRUE
+          $.cge
+  memory > i
+    FALSE
+  cge > @
+
+```
```

Comment:
@maxonfjvipon this is more illustrating example, not working code,

Reply:
@levBagryansky why doesn't it work? Everything seems fine

Reply:
@maxonfjvipon I get an error EOcorrect_cage_recursionTest.works:124->PhDefault.attr:250->PhDefault.attr:248->PhDefault.attr:250 » Ex D="Error at "EOorg.EOeolang.EObytes$EOas_bool#φ" attribute; caused by Can't #copy() package object 'bytes'" here and don't know why